File: solutions/Leetcode_1696/leetcode_1696.cc
#include <deque>
#include <queue>
#include <vector>

using namespace std;

class Solution1
{
public:
    int maxResult(vector<int> &nums, int k)
    {
        deque<pair<int, int>> q;
        q.push_back({0, nums[0]});
        int ret = nums[0];
        for (int i = 1; i < nums.size(); i++)
        {
            while (!q.empty() && q.back().first < (i - k))
                q.pop_back();
            ret = q.back().second + nums[i];
            while (!q.empty() && q.front().second < ret)
                q.pop_front();
            q.push_front({i, ret});
        }
        return ret;
    }
};

class Solition2
{
public:
    int maxResult(vector<int> &nums, int k)
    {
        priority_queue<pair<int, int>> p;
        p.push({nums[0], 0});
        int ret = nums[0];
        for (int i = 1; i < nums.size(); i++)
        {
            while (!p.empty() && p.top().second < (i - k))
                p.pop();
            ret = p.top().first + nums[i];
            p.push({ret, i});
        }
        return ret;
    }
};

File: solutions/Leetcode_1696/leetcode_1696.py
import heapq
from collections import deque
from typing import List


class Solution1:
    def max_result(self, nums: List[int], k: int) -> int:
        q = deque(([(0, nums[0])]))
        ret = nums[0]
        for i in range(1, len(nums)):
            while q and q[-1][0] < (i - k):
                q.pop()
            ret = nums[i] + q[-1][1]
            while q and q[0][1] < ret:
                q.popleft()
            q.appendleft((i, ret))
        return ret


class Solution2:
    def max_result(self, nums, k):
        h = [(-nums[0], 0)]
        ret = nums[0]
        for i in range(1, len(nums)):
            while h and h[0][1] < (i-k):
                heapq.heappop(h)
            ret = -h[0][0] + nums[i]
            heapq.heappush(h, (-ret, i))
        return ret
